portal-bridge/src/bridge/state.rs

```diff
@@ -0,0 +1,226 @@
+use std::{path::PathBuf, sync::Arc};
+
+use anyhow::{anyhow, ensure};
+use scraper::{Html, Selector};
+use tokio::{
+    sync::{OwnedSemaphorePermit, Semaphore},
+    task::JoinHandle,
+    time::timeout,
+};
+use tracing::{debug, error, info, warn};
+use trin_metrics::bridge::BridgeMetricsReporter;
+
+use crate::{
+    bridge::history::{EPOCH_SIZE, SERVE_BLOCK_TIMEOUT},
+    gossip::gossip_state_content,
+    types::{
+        era1::Era1,
+        mode::{BridgeMode, ModeType},
+        state::{
+            content::{create_content_key, create_content_value},
+            execution::State,
+        },
+    },
+};
+use ethportal_api::{jsonrpsee::http_client::HttpClient, StateContentKey, StateContentValue};
+use surf::{Client, Config};
+
+use trin_validation::oracle::HeaderOracle;
+
+const ERA1_DIR_URL: &str = "https://era1.ethportal.net/";
+const ERA1_FILE_COUNT: usize = 1897;
+
+pub struct StateBridge {
+    pub mode: BridgeMode,
+    pub portal_clients: Vec<HttpClient>,
+    pub header_oracle: HeaderOracle,
+    pub epoch_acc_path: PathBuf,
+    pub era1_files: Vec<String>,
+    pub http_client: Client,
+    pub metrics: BridgeMetricsReporter,
+    pub gossip_limit: usize,
+}
+
+impl StateBridge {
+    pub async fn new(
+        mode: BridgeMode,
+        portal_clients: Vec<HttpClient>,
+        header_oracle: HeaderOracle,
+        epoch_acc_path: PathBuf,
+        gossip_limit: usize,
+    ) -> anyhow::Result<Self> {
+        let http_client: Client = Config::new()
+            .add_header("Content-Type", "application/xml")
+            .expect("to be able to add header")
+            .try_into()?;
+        let era1_files = get_list_of_era1_files(&http_client).await?;
+        let metrics = BridgeMetricsReporter::new("state".to_string(), &format!("{mode:?}"));
+        Ok(Self {
+            mode,
+            portal_clients,
+            header_oracle,
+            epoch_acc_path,
+            era1_files,
+            http_client,
+            metrics,
+            gossip_limit,
+        })
+    }
+
+    pub async fn launch(&self) {
+        info!("Launching state bridge: {:?}", self.mode);
+        match self.mode.clone() {
+            BridgeMode::Single(ModeType::Block(last_block)) => self.launch_state(last_block).await,
+            _ => panic!("State bridge only supports State modes."),
```

Comment:
I am assuming you mean the if latest_block > X, this check will be removed in the next PR so it shouldn't be too much of a concern

Reply:
No, I meant entire mode check. Why not store only `last_block` in self instead of `mode`?
But if you plan to refactor this again in the next PR, it's fine to leave it here.

Reply:
I do plan to refactor it yeah. Eventually I want to add a snapshot mode

Reply:
I was confused thank you for clearing my confusion for me!